sdks/python/apache_beam/runners/portability/local_job_service.py

```diff
@@ -185,55 +152,129 @@ class BeamJob(threading.Thread):
     self.state = beam_job_api_pb2.JobState.STARTING
-    self.daemon = True
-
-  def add_state_change_callback(self, f):
-    self._state_change_callbacks.append(f)
-    f(self.state)
-
-  def add_log_callback(self, f):
-    self._log_callbacks.append(f)
 
   @property
   def state(self):
     return self._state
 
   @state.setter
   def state(self, new_state):
-    for state_change_callback in self._state_change_callbacks:
-      state_change_callback(new_state)
+    """Sets the job state.
+
+    This will inform GetStateStream and GetMessageStream of the new state.
+    """
+
+    self._state_notifier.acquire()
+    self._log_notifier.acquire()
+
+    self._state_changes.append(new_state)
     self._state = new_state
 
+    self._log_notifier.notify_all()
+    self._state_notifier.notify_all()
+
+    self._log_notifier.release()
+    self._state_notifier.release()
+
+  def write_log(self, log):
+    self._log_notifier.acquire()
+    self._logs.append(log)
+    self._log_notifier.notify_all()
+    self._log_notifier.release()
+
+  def _cleanup(self):
+    self._state_notifier.acquire()
+    self._log_notifier.acquire()
+
+    self._final_log_count = len(self._logs)
+    self._final_state_count = len(self._state_changes)
+
+    self._state_notifier.notify_all()
+    self._log_notifier.notify_all()
+
+    self._log_notifier.release()
+    self._state_notifier.release()
+
   def run(self):
-    with JobLogHandler(self._log_callbacks):
+    with JobLogHandler(self):
       try:
         fn_api_runner.FnApiRunner().run_via_runner_api(self._pipeline_proto)
         logging.info('Successfully completed job.')
         self.state = beam_job_api_pb2.JobState.DONE
       except:  # pylint: disable=bare-except
         logging.exception('Error running pipeline.')
-        traceback.print_exc()
+        logging.exception(traceback.format_exc())
         self.state = beam_job_api_pb2.JobState.FAILED
         raise
+      finally:
+        # In order for consumers to read all messages, this must be the final
+        # instruction after a terminal state.
+        self._cleanup()
 
   def cancel(self):
     if self.state not in TERMINAL_STATES:
       self.state = beam_job_api_pb2.JobState.CANCELLING
       # TODO(robertwb): Actually cancel...
       self.state = beam_job_api_pb2.JobState.CANCELLED
+      self._cleanup()
+
+  def _stream_array(self, array, notifier, final_count):
+    """Yields all elements in array until array length reaches final_count.
+
+    This method streams all elements in array. It uses the notifier to wait
+    until new messages are received. The streams ends when this method emits
+    up to final_count number of elements.
+    """
+    index = 0
+
+    # Pull all  changes until the job finishes.
+    while index != self.__dict__[final_count]:
+      notifier.acquire()
+      notifier.wait(5)
+      while index < len(array):
+        yield array[index]
+        index += 1
+      notifier.release()
+
+  def GetStateStream(self):
```

Comment:
Only GRPC methods should be CamelCase, otherwise follow PEP 8.

Reply:
Sounds good, changed to PEP 8